Docs/604 use submenu interaction (#765)
pluralsight/tva: website/docs/development/react-utils/use-submenu-interaction.mdx

---

> tags: [Development, Packages, React, React utils, useSubmenuInteraction, hooks]

Reviewer: Brad made a good point that the search should automatically tag these since it's the page title, etc.

```suggestion
tags: [Development, Packages, React, hooks]
```

title: useSubmenuInteraction
---

import CodeBlock from '@theme/CodeBlock'
import ViewSourceLink from '@site/src/components/ViewSourceLink/ViewSourceLink'

A custom hook for handling submenu interactions and state within a [menu](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menu_role) role.

<ViewSourceLink
  icon="github"
  href="https://github.com/pluralsight/tva/blob/main/packages/react-utils/src/hooks/menu/useSubmenuInteraction.ts"
>
  View source
</ViewSourceLink>

## Import

```javascript
import { useSubmenuInteraction } from '@pluralsight/react-utils'
```

## Basic usage

The `useSubmenuInteraction` hook takes no arguments.

```jsx
const { expanded, menu, trigger } = useMenuInteraction()
```

## Return value

> The `useSubmenuInteraction` hook returns prop sets for the submenu and its trigger, as well as a boolean indicating the `expanded` state of the submenu.

Reviewer: ```suggestion
The `useSubmenuInteraction` hook returns props for the submenu and its trigger and the `expanded` state of the submenu.
```

```typescript

> interface useSubmenuInteractionProps {

Reviewer: interfaces are supposed to use a Capitalized naming convention. 😉 

```suggestion
interface SubmenuInteractionProps {
```

  expanded: boolean
  menu: {
    'aria-expanded': 'true' | 'false'
    ref: RefObject<HTMLMenuElement>
    role: 'menu'
    onBlur: (event: FocusEvent) => void
    onKeyDown: (event: KeyboardEvent) => void
  }
  trigger: {

>     'data-expanded': 'true' | 'false'

Reviewer: ```suggestion
    'data-expanded': boolean
```

    ref: RefObject<HTMLButtonElement>
    onClick: () => void
    onKeyDown: (event: KeyboardEvent) => void
  }
}
```

## Example usage in component

Here is an example of how you might use this hook in a Menu component combined with the [Menu headless-styles](../headless-styles/Menu.mdx) and [useRovingTabIndex](./use-roving-tabindex.mdx) hook for enhanced UX and accessibility.

```jsx
export function MenuItem(props) {
  const rovingTabIndexProps = useRovingTabIndex()
  const menuStyles = getMenuProps()

  return (
    <li {...menuStyles.menuListItem}>
      <a {...menuStyles.menuItem} {...rovingTabIndexProps} href={props.href}>
        {props.children}
      </a>
    </li>
  )
}

export function Submenu(props) {
  const submenuStyles = getMenuProps({
    kind: 'submenu',
    label: props.label,
  })

>   const { menu, trigger } = useSubmenuInteraction()

Reviewer: ```suggestion
 // higlight-next-line
  const { menu, trigger } = useSubmenuInteraction()
```
